### ISSUE TITLE
pickup and delivery option with cspy

### ISSUE THREAD

**Issue by Kuifje02** [Kuifje02/vrpy#15]
Just had a quick google about the pickup and delivery and found this [paper](https://backend.orbit.dtu.dk/ws/portalfiles/portal/4128533/PDPTW-BCP-TS-v3.pdf).
End of page 10 basically says that the bidirectional algorithm won't work due to dominance conditions for backward labels. 
Might still be worth running these with `direction="forward"` after we figure out how to model it

_Originally posted by @torressa in https://github.com/Kuifje02/vrpy/issues/14#issuecomment-622475443_

**Comment by torressa** [comment]
Had a look last weekend, might be harder than I thought.
The check in the REF is pretty easy, just needs the current partial path (already done in cspy/dev) but the algorithm has to allow for paths to revisit nodes. I think it should, at least in theory, but definitely doesn't now

**Comment by Kuifje02** [comment]
Why should the algorithm allow to revisit nodes ? Is it a modeling prerequisite, or is it just potentialy too time consuming to have elementary paths?

**Comment by torressa** [comment]
You're right, I thought it was a modelling thing. Re-read the paper, can actually solve it with elementary paths... For which we should be alright!
So, an additional one or two resources that check whether the delivery node has already been visited should be fine for this.
Will work on it after the new cspy release (nearly there!)

**Comment by shunana** [comment]
HI
I wanted to ask, how far are you with this ? It is something that I would like to use, maybe we could collaborate on making it work?

Thanks

**Comment by Kuifje02** [comment]
Hi @shunana, currently we are not working on this (lack of time), this is probably quite a bit of work, and any help / collaboration would be great. I think the hardest part will be implementing the pickup and delivery constraints in cspy, with a resource logic.

**Comment by torressa** [comment]
Sorry, completely forgot about this!
I'll push an attempt, seems to give a feasible solution for the `examples/pdp.py`, doesn't match the objective value though
From the paper, I've added an attempt for the `open_requests` check in the callback, the unreachable nodes check is already in built in cspy. The only thing that doesn't match is the dominance check, but shouldn't affect I don't think...

**Comment by shunana** [comment]
Great @Kuifje02 ! Okay I will look at it and see how it will go. Nice that you already have had a look at that @torressa , I will also have a look on that.

**Comment by torressa** [comment]
Nice one! 
I think the REF does what we want, just a few things left on the cspy side. 
Following the paper, the dominance check for the load (resource index 2) should be disabled (as it can be negative), this is pretty straightforward to do in cspy (not through the user interface but in the code). 
Also, need to look into the transformation of the pricing coefficients. I think this may be what is causing issues but I'm not sure.

**Comment by torressa** [comment]
Realised it was mostly cspy stuff. So the current version uses some the cspy the [vrpy_patch15](https://github.com/torressa/cspy/tree/vrpy_patch15) branch. Had to include the open request subset comparison when performing dominance checks. This is really expensive
Gives the optimal solution.. but its slower than the LP! So still plenty of work to do!

**Comment by torressa** [comment]
I've updated the implementation (on the cspy side) to what should be OK now (updating open requests sets on the fly, as opposed the previous terribly inefficient prototype).
The optimal solution is reached much faster than the LP, however it takes a very long time to converge, making it actually slower. 

Not sure how to proceed here, the paper suggests adding several inequalities (pickup-delivery related) to strengthen the formulation and probably also help with convergence. This maybe raises the need to implement the generic cutting framework so we can add all try out new and varied inequalities, maybe non-robust ones (which add resources to the subproblems).

**Comment by torressa** [comment]
Had another go at this. I'm playing around with allowing non-elementary routes (will submit a PR soon) and it works much better (at least better than the LP)! 

The code for this (PDP) is very specific to VRPy, so not sure what the best approach would be here. We could move the VRPY-specific version of cspy out as a separate repo or maybe even here. With this we could go further into moving more subproblem parts into C++, like writing the REF functions (after passing the relevant data).
